torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
 
         else:
             raise NotImplementedError("Method not yet implemented")
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
+def first_one(n: Node):
+    res = []
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        res = [Equality(arg_type.__args__[0], n.type.__args__[0])]
+    return res
+
+# todo needs review for addition. Is this constraint correct?
+@register_refinement_rule(BatchNorm2d)
+@register_refinement_rule(torch.nn.ReLU)
+@register_refinement_rule(torch.nn.AdaptiveAvgPool2d)
+def all_eq(n: Node):
+    res = []
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        args1 = arg_type.__args__
+        args2 = n.type.__args__
+        res = [Equality(args1[i], args2[i]) for i in range(len(args1))]
+    return res
+
+@register_refinement_rule(torch.add)
+@register_refinement_rule(operator.add)
+def add_eq(n: Node):
+    res = []
+    if isinstance(n.args[0], Node) and isinstance(n.args[1], Node):
+        arg_type1 = n.args[0].type
+        arg_type2 = n.args[1].type
+        if isinstance(arg_type1, TensorType) and isinstance(arg_type2, TensorType) and isinstance(n.type, TensorType):
+            args1 = arg_type1.__args__
+            args2 = arg_type2.__args__
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
+                if x == y:
```

Comment:
I think it would be easy enough to add a check for the `1`  behavior of broadcasting

Reply:
this part is handled in the type checker already! the definitions are here https://github.com/pytorch/pytorch/blob/faf82478fec1d777f1120ee2d1e06f01a1bf3325/torch/fx/experimental/graph_gradual_typechecker.py#L45

by the time we reach the symbolic inference part, we already did the best we can with any concrete types we had earlier.

Reply:
i guess the case i'm talking about is if one value is `1` the output will be the symbolic shape of the other

Reply:
I added a test case for this here https://github.com/pytorch/pytorch/pull/61776/files#diff-8ccda303dbd4c687535083b21b946c204650cb45561330e48b56a2299ab92752R922

is that along the lines of what you're referring to?